src/InputState.h

```diff
@@ -231,6 +241,92 @@ struct AssociatedPhrasesPlain : InputState {
+      return "[中文數字] " + number;
+    } else if (style == ChineseNumberStyle::UPPER) {
+      return "[大寫數字] " + number;
+    } else if (style == ChineseNumberStyle::SUZHOU) {
+      return "[蘇州碼] " + number;
+    }
+    return number;
+  }
+
+  std::string number;
+  ChineseNumberStyle style;
+};
+
+static std::string DateMacros[] = {"MACRO@DATE_TODAY_SHORT",
+                                   "MACRO@DATE_TODAY_MEDIUM",
+                                   "MACRO@DATE_TODAY_MEDIUM_ROC",
+                                   "MACRO@DATE_TODAY_MEDIUM_CHINESE",
+                                   "MACRO@DATE_TODAY_MEDIUM_JAPANESE",
+                                   "MACRO@THIS_YEAR_PLAIN",
+                                   "MACRO@THIS_YEAR_PLAIN_WITH_ERA",
+                                   "MACRO@THIS_YEAR_ROC",
+                                   "MACRO@THIS_YEAR_JAPANESE",
+                                   "MACRO@DATE_TODAY_WEEKDAY_SHORT",
+                                   "MACRO@DATE_TODAY_WEEKDAY",
+                                   "MACRO@DATE_TODAY2_WEEKDAY",
+                                   "MACRO@DATE_TODAY_WEEKDAY_JAPANESE",
+                                   "MACRO@TIME_NOW_SHORT",
+                                   "MACRO@TIME_NOW_MEDIUM",
+                                   "MACRO@THIS_YEAR_GANZHI",
+                                   "MACRO@THIS_YEAR_CHINESE_ZODIAC"};
+
+struct SelectingDateMacro : InputState {
+  SelectingDateMacro(std::function<std::string(std::string)> converter)
+      : converter(converter) {
+    for (std::string macro : DateMacros) {
+      menu.emplace_back(converter(macro));
+    }
+  }
+
+  std::vector<std::string> menu;
+  std::function<std::string(std::string)> converter;
+};
+
+struct SelectingFeature : InputState {
+  struct Feature {
+    Feature(std::string name,
+            std::function<std::unique_ptr<InputState>(void)> nextState)
+        : name(std::move(name)), nextState(std::move(nextState)) {}
+    std::string name;
+    std::function<std::unique_ptr<InputState>(void)> nextState;
+  };
+
+  SelectingFeature(std::function<std::string(std::string)> converter)
+      : converter(converter) {
+    features.emplace_back(std::make_unique<Feature>("日期與時間", [this]() {
+      return std::make_unique<SelectingDateMacro>(this->converter);
+    }));
+    features.emplace_back(std::make_unique<Feature>("中文數字", []() {
+      return std::make_unique<ChineseNumber>("", ChineseNumberStyle::LOWER);
+    }));
+    features.emplace_back(std::make_unique<Feature>("大寫數字", []() {
+      return std::make_unique<ChineseNumber>("", ChineseNumberStyle::UPPER);
+    }));
+    features.emplace_back(std::make_unique<Feature>("蘇州碼", []() {
+      return std::make_unique<ChineseNumber>("", ChineseNumberStyle::SUZHOU);
+    }));
+  }
+
+  std::unique_ptr<InputState> nextState(size_t index) {
+    return features[index]->nextState();
+  }
+
+  std::vector<std::unique_ptr<Feature>> features;
```

Comment:
Since `Feature`s are simply read during an enumeration, I don't think you need to use `std::unique_ptr`s.

Reply:
Fixed in 92e46601d055fb57699128d5d2cb8634a1fef339